### PR TITLE
Issue11 advanced load of config file

### LINKED ISSUE #11: Cleanly manage configuration keys which require environment variables
Some configuration keys contain the name of an environment variable to be used to retrieve the value (typically for passwords).
These keys are not differentiated from "normal" keys in the configuration, and must be processed differently in the harvester's code.
It would be better to retrieve the value when parsing the configuration file and process all configuration keys the same way.

### PR DESCRIPTION
**Pull request by opsdep**
closes #11 
The passwords are being read carefully _without involvement of harvesters code_. It is done the **config reading** part of the code as requested in the issue description. One more test is added for this reason.

### CODE REVIEW

**Review by aperrin66 — CHANGES_REQUESTED** [review]
- The `_validate` method should only validate the data, not modify it.
- This solution works only for a "password" field. We want something that works for any environment variable.

Please have a look at the "Constructors, representers, resolvers" section in the [PyYAML doc](https://pyyaml.org/wiki/PyYAMLDocumentation) in order to define a custom tag for environment variables.
[This](https://stackoverflow.com/questions/43058050/creating-custom-tag-in-pyyaml) can also help.

**Comment by opsdep** [comment]
@aperrin66 
All addressed in e3305a3. The file of the test shows how should we use the tag inside the yml file.

**Comment by opsdep** [comment]
@aperrin66 
The extra code is deleted in bd09b6f. This is done for better coverage percentage.

File: geospaas_harvesting/harvest.py [hunk 1]
```diff
@@ -60,9 +59,20 @@ def _validate(self):
             assert self.HARVESTER_CLASS_KEY in config.keys(), (
                 "Harvester configuration must contain the following keys: " +
                 ', '.join(self.HARVESTER_CLASS_KEY))
-
             assert isinstance(config['class'], str), (
                 f"In '{name}' section: 'class' must be a string")
+            if 'password' in config:
+                assert config['password'] != None, "An environmental variable (not a None) is needed\
+                that has been previousely set with the same name in front of 'password' section of \
+                    configuration of this harvester"
```
Comment (aperrin66) [inline geospaas_harvesting/harvest.py]: We either need a generic way to validate harvesters parameters, of leave it as it is and rely on the exceptions raised in the rest of code.

Also, please avoid breaking lines using `\` as much as possible (see [PEP8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length)).

File: geospaas_harvesting/harvest.py [hunk 2]
```diff
@@ -60,9 +59,20 @@ def _validate(self):
...
+
+    class EnvTag(yaml.YAMLObject):
+        """class for reading the tags of yml file for finding the value of environmental variables"""
```
Comment (aperrin66) [inline geospaas_harvesting/harvest.py]:
```suggestion
        """class for reading the tags of yml file for finding the value of environment variables"""
```

File: tests/test_daemon_script.py [hunk 3]
```diff
@@ -49,6 +50,27 @@ def test_loading_valid_conf(self):
             }
         )
 
+    def test_loading_valid_conf_with_password(self):
```
Comment (aperrin66) [inline tests/test_daemon_script.py]: Small remark about naming: what you did is not only for passwords, but for any configuration parameter containing an environment variable.

**Review by aperrin66 — CHANGES_REQUESTED** [review]
That's much better! Almost there.
Please check for PEP8 conformity.

File: geospaas_harvesting/harvest.py (hunk 2 above)
Comment (opsdep) [reply geospaas_harvesting/harvest.py]: Done in the next commit

File: tests/test_daemon_script.py (hunk 3 above)
Comment (opsdep) [reply tests/test_daemon_script.py]: Done in the next commit

File: geospaas_harvesting/harvest.py (hunk 1 above)
Comment (opsdep) [reply geospaas_harvesting/harvest.py]: That was a very long string which can't we separated into several line unless we use the `\` or we use the `"""` for it. Done in the next commit.

**Comment by opsdep** [comment]
all addressed in c37a1fe.

File: geospaas_harvesting/harvest.py [hunk 4]
```diff
@@ -60,9 +59,20 @@ def _validate(self):
             assert self.HARVESTER_CLASS_KEY in config.keys(), (
                 "Harvester configuration must contain the following keys: " +
                 ', '.join(self.HARVESTER_CLASS_KEY))
-
             assert isinstance(config['class'], str), (
                 f"In '{name}' section: 'class' must be a string")
+            if 'password' in config:
+                assert config['password'] != None, """An environment variable (not a None) is needed
+                 that has been previousely set with the same name in front of 'password' section of
+                 configuration of this harvester"""
```
Comment (aperrin66) [inline geospaas_harvesting/harvest.py]: This previous comment hasn't been taken into account:
> We either need a generic way to validate harvesters parameters, of leave it as it is and rely on the exceptions raised in the rest of code.

**Comment by opsdep** [comment]
@aperrin66 
All done in a845c30.

**Review by aperrin66 — APPROVED** [review]

**Pull request merged**